pkg/dnscache/dnscache.go

```diff
@@ -0,0 +1,223 @@
+
+	return queryResult, nil
+}
+
+// addRootNode adds a new root node for DNS queries
+func (nc *DNSCache) addRootNode(dns *trace.ProtoDNS, timestamp time.Time) {
+	nodeType := DNS
+	value := dns.Questions[0].Name
+	if dns.Questions[0].Type == "PTR" {
+		// handle reverse query case
+		nodeType = IP
+		// trim whatever suffix is appended (can't know which ip version solely from PTR record)
+		value = strings.TrimSuffix(value, reverseQueryIPv4Suffix)
+		value = strings.TrimSuffix(value, reverseQueryIPv6Suffix)
+	}
+
+	// maximum value for time.Time in go
+	// this is used so root nodes will never expire
+	maxTime := time.Unix(1<<63-62135596801, 999999999)
+
+	// build the root node
+	node := &dnsNode{
+		value:        value,
+		nodeType:     nodeType,
+		next:         newNodeSet(),
+		parents:      newNodeSet(),
+		expiredAfter: maxTime,
+	}
+
+	// index it
+	nc.queryIndices[node.value] = node
+	nc.queryRoots.Add(node.value, node)
+	// build and index its child nodes
+	nc.addChildNodes(dns.Answers, node, timestamp)
+}
+
+// addChildNodes builds new DNS nodes from a ProtoDNS object and attaches them to a parent node
+func (nc *DNSCache) addChildNodes(answers []trace.ProtoDNSResourceRecord, parent *dnsNode, timestamp time.Time) {
+	for _, answer := range answers {
+		// Check if there is another parent node based on the answer name.
+		// For example in A/AAAA queries the answers may include first the CNAME,
+		// and then actual IP records related to the CNAME in the name.
+		// If the answer's name isn't checked then we may wrongly relate to the original
+		// query and not the CNAME.
+		contextParent := parent
+		actualParentName := answer.Name
+		if contextParent.value != actualParentName {
+			actualParentNode, ok := nc.queryIndices[actualParentName]
+			if ok {
+				// switch parent if possible, otherwise stick with the original
+				contextParent = actualParentNode
+			}
+		}
+
+		// build a node from the answer and discovered parent
+		// (it may be discarded if the node was previously indexed)
+		node := makeNodeFromAnswer(contextParent, &answer, timestamp)
+
+		// check if the child node was already indexed
+		if child, ok := nc.queryIndices[node.value]; !ok {
+			// index if not...
+			nc.queryIndices[node.value] = node
+			// ...and add the node
+			contextParent.next.Append(node)
+		} else {
+			// if already indexed register the parent in the child...
+			child.parents.Append(contextParent)
+			// ...update the TTL...
+			child.updateTTL(answer.TTL, timestamp)
+			// ...and register the existing child to the parent
+			contextParent.next.Append(child)
+		}
+	}
+}
+
+// evictQueryRoot is the eviction function for the query roots LRU
+func (nc *DNSCache) evictQueryRoot(addr string, node *dnsNode) {
+	// eviction may sometimes occur in the context of an Add transaction.
+	// In these cases locking is unnecessary, since the evict is done in
+	// sequence with adding.
+	locked := nc.lock.TryLock()
+	if locked {
```

Comment:
So this is unlikely because essentially two cases are possible in the current code (and only one can actually be triggered - the 2nd one):
1. A manual eviction is started, in this case the eviction acquires the lock before any other tx or waits for it. We don't have code manually calling eviction right now.
2. This happens in the context of adding an element to the LRU, this can only happen in the context of the tx mutex already being locked, so the TryLock fails and skips the defer.